### ISSUE TITLE
No/broken kerning for a large number of fonts

### ISSUE THREAD

**Issue by birdie-github** [harfbuzz/harfbuzz#2394]
I really don't know what to do 'cause Pango/RH developers are unwilling to cooperate or even admit the issue, so it's my last resort.

The pertinent details are here: https://gitlab.gnome.org/GNOME/pango/issues/463
The original bug report is here: https://bugzilla.redhat.com/show_bug.cgi?id=1767499

It looks like the bug is directly related to harfbuzz since the regression range indicates that it's due to the switch from FreeType to HarfBuzz that the issue first occurred.

**Comment by ebraminio** [comment]
Hi there, am sorry about the path you went, thanks for following the issue all along and hopefully you will have a little better experience here :)

What here can be very helpful is to post an exact testcase to show the issue, there is a tool distributed along harfbuzz, in Debian in a separate package called libharfbuzz-bin, `hb-view`, that can make debugging of the issue easier from our side, since you suspect it is because of Pango font loading migration from FreeType to HarfBuzz, we will pass --font-funcs to investigate the difference, so what we need is a command like:

    hb-view FULL-PATH-TO-YOUR-TTF-FILE.ttf "some text" --font-size=FONTSIZE --font-funcs=ot | display

And its comparison with:

    hb-view FULL-PATH-TO-YOUR-TTF-FILE.ttf "some text" --font-size=FONTSIZE --font-funcs=ft | display

`display` command is distributed by imagemagick package. Replace FONTSIZE with a number you think will show the issue, and please share the font with us somehow, possibly using the email I put in my GitHub profile, and if you think these are too technical for you or anything, we will find some other way to sort this out possibly using some IM platform :) thanks!

**Comment by ebraminio** [comment]
Oh, hmmm, so this is hinting related now I see, guess you should ignore my previous comment :/

`hb-view tahoma.ttf "Run Program" --font-funcs=ot --font-size=12 | display`

![image](https://user-images.githubusercontent.com/833473/81475670-ebc01980-9222-11ea-8124-a7a8c3e71015.png)

**Comment by ebraminio** [comment]
IIUC this issue is the same as #892.

**Comment by birdie-github** [comment]
I'm kinda lost in all of this.

What does it leave us with? Is this an issue with HarfBuzz? Pango? Cairo? Something else? Are there any patches which I could apply on top of these packages to fix the issue which pretty much breaks kerning altogether?

The bug report you're referring to is closed as WONTFIX.

Matthias Clasen seemingly couldn't care less about the bug report I filed against Pango.

This is all kinda ... unpleasant to put it mildly. It's now been almost a year since the issue arose.

**Comment by behdad** [comment]
> Matthias Clasen seemingly couldn't care less about the bug report I filed against Pango.
> 
> This is all kinda ... unpleasant to put it mildly. It's now been almost a year since the issue arose.

We decided to not support aggressive hinting in Pango moving forward.  What you want is unsupported by Pango moving forward.  That's why you are getting Matthias and others ignoring you.  That's all.

**Comment by birdie-github** [comment]
> What you want is unsupported by Pango moving forward.

This doesn't ... look right. I mean you're talking about certain fonts families not being correctly rendered by this stack (Pango/Harfbuzz/Cairo).

So, what will we have moving forward? I'm utterly confused. We had fonts being rendered more or less correctly by Pango 1.43 which used FreeType and now in 1.44 ... we have something brand new ... which doesn't work in many situations? My head is about to implode.

**Comment by ebraminio** [comment]
> I mean you're talking about certain fonts families not being correctly rendered by this stack

I think those certain fonts families are rendered correctly but not if you pass https://gitlab.gnome.org/GNOME/pango/uploads/5c01170e69910a8a3f0dd2c1b72b5122/fonts.conf to your `.fonts.conf`.

**Comment by birdie-github** [comment]
May I ask what's wrong's with this file? It gives me the absolute best results in terms of font rendering on my 96 DPI screen. I've tried other hinting styles but only `hintfull` satisfies me as it looks very close to ClearType v2 in Windows 7 which in my opinion has the best font rendering out of any implementations.

And I'm not rich enough to afford a retina/HiDPI display (in which case I could have disabled hinting altogether). Is this a new requirement for the _modern_ Linux font stack?

**Comment by birdie-github** [comment]
If my .fonts.conf is "broken", "incorrect" or "outdated" please tell me how I can achieve _the same results with the new font stack_.

**Comment by behdad** [comment]
> I've tried other hinting styles but only `hintfull` satisfies me as it looks very close to ClearType v2 in Windows 7 which in my opinion has the best font rendering out of any implementations.

Exactly.  That's what we decided not to support anymore.  You can try to get used to the "blurry" rendering.  Or you can find something else to use, or whatever.  You are using Free Software you realize, right?

**Comment by behdad** [comment]
> If my .fonts.conf is "broken", "incorrect" or "outdated" please tell me how I can achieve _the same results with the new font stack_.

You can't.

**Comment by ebraminio** [comment]
> [..] please tell me how I can achieve the same results with the new font stack.

Maybe it isn't possible for some reason? As you noticed even modern browsers on Windows can't offer that experience anymore as they value subpixel positioning.

I understand what you want, I really do as I really wanted to implement such kind of hinting for HarfBuzz and of course that has less things to do with Pango's decision, I don't think however calling entire [mostly voluntarily developed] text stack broken just because it doesn't provide a specific need for some explained technical reasons is a nice idea.

**Comment by ebraminio** [comment]
Closing this as its decision doesn't belong to HarfBuzz.

However this has motivated me more to implement such kind of hinting maybe in HarfBuzz that possibly Pango wants to use someday.

Thanks

**Comment by mattst88** [comment]
> > Matthias Clasen seemingly couldn't care less about the bug report I filed against Pango.
> > This is all kinda ... unpleasant to put it mildly. It's now been almost a year since the issue arose.
> 
> We decided to not support aggressive hinting in Pango moving forward. What you want is unsupported by Pango moving forward. That's why you are getting Matthias and others ignoring you. That's all.

Is there something written about that decision so people can understand? A mailing list thread maybe?

**Comment by ebraminio** [comment]
https://blogs.gnome.org/mclasen/2019/05/25/pango-future-directions/ AFAIK

**Comment by liskin** [comment]
It's worth noting that the vertical grid alignment shown in this picture: https://user-images.githubusercontent.com/4958603/81482131-49a13100-9224-11ea-9d9b-a8362ac059a4.png hasn't worked in some browsers for a while now on my machine, but having hintfull and `FREETYPE_PROPERTIES=truetype:interpreter-version=35` still results in pretty sharp a readable text:

![2020-05-09-215518_751x218_scrot](https://user-images.githubusercontent.com/300342/81483630-ecb37400-923f-11ea-912d-ce20bb523303.png)

When I upgrade to pango 1.44, the browsers still work the same, possibly because they use harfbuzz and then do the subpixel rendering outside of pango? But all the normal gtk applications are completely broken. So there really is a bug somewhere that should be fixed. Perhaps the distros just need to upgrade cairo to the newer one that supports subpixel positioning?

**Comment by liskin** [comment]
(And I'd also like to express my disappointment with the decision not to support hintfull, as that means people with above average visual acuity will be unable to use gtk without a HiDPI display, whereas they were until now. And I think there's not enough cheap HiDPI displays out there, and there's not enough software support for them either. It's just too soon. But I fully understand that if this work is being done by volunteers only, there isn't all that much we can do about it. It's a sad day for people with good eyes and deep pockets.)

**Comment by behdad** [comment]
> It's just too soon.

When would be appropriate?  2030?  I'm sure someone would complain then as well.  The longtail never ends.  That's fact of life.  Use whatever works for you.

**Comment by liskin** [comment]
> When would be appropriate?

Well, I'd say the year when most laptops sold have hidpi screen might be a good first guess. Right now some manufacturers are still selling 1366x768 screens, and most laptops have fullhd. In 10 years I'm fairly certain this will change.

Yeah, I understand that you've had a hidpi screen for a while and supporting lower resolutions seems boring. But it really is used by an overwhelming majority of Linux users... :-/

**Comment by GitHubIsToxic** [comment]
I have never even encountered a HiDPI display in the wild yet, certainly none that I can afford. I just bought two new displays last year (max res 1920x1080), and I expect to use them for the next 5 years.

Planned obsolescence in hardware is one thing, but through software!?

**Comment by ebraminio** [comment]
> I understand that you've had a hidpi screen for a while and supporting lower resolutions seems boring.

This feels like [ad hominem](https://en.wikipedia.org/wiki/Ad_hominem) kind of argument instead focusing on not personal arguments like such kind of rendering undermines subpixel positioning attempts and browsers also opted out from it.

**Comment by mattst88** [comment]
> > I've tried other hinting styles but only `hintfull` satisfies me as it looks very close to ClearType v2 in Windows 7 which in my opinion has the best font rendering out of any implementations.
> 
> Exactly. That's what we decided not to support anymore. You can try to get used to the "blurry" rendering. Or you can find something else to use, or whatever. You are using Free Software you realize, right?

I don't want to tell you how to act or what you should do, but as a fellow free software developer and volunteer that not-infrequently has to deal with angry users, I find it much more productive to acknowledge the users problem even if you are not going to fix it. I prefer to just say, "Yeah, I know this isn't ideal and I'm sorry but for reasons X, Y, Z we decided that this was an acceptable trade off." Most users won't necessarily be happy but will accept this.

**Comment by liskin** [comment]
@ebraminio That was actually an attempt to empathise with Behdad. :-( I'm sorry if it came across as ad hominem.

**Comment by khaledhosny** [comment]
HarfBuzz is not a font renderer and it does not do hinting by designed, if Pango needs hinted advances HarfBuzz is not where it can get them from. This issue should be closed, the discussion is completely off-topic to HarfBuzz.

**Comment by birdie-github** [comment]
> This issue should be closed, the discussion is completely off-topic to HarfBuzz.

Pango developers so far have remained mostly silent: https://gitlab.gnome.org/GNOME/pango/issues/463 and there are just two things which put fonts on the screen in GTK apps: Pango and HarfBuzz. FreeType does some work as well but it's not fully utilized any longer. What does this leave us with? I'm confused.

**Comment by markg85** [comment]
> > I've tried other hinting styles but only `hintfull` satisfies me as it looks very close to ClearType v2 in Windows 7 which in my opinion has the best font rendering out of any implementations.
> 
> Exactly. That's what we decided not to support anymore. You can try to get used to the "blurry" rendering. Or you can find something else to use, or whatever. You are using Free Software you realize, right?

That comment is a blunt slap in the face in a "polite" way. You fully *realize* that a fix for the topic starter would probably mean forking if he persists to using the same fonts. And keeping that alive and actually working is a very big time consuming "hobby" that one quickly loses interest of.

Instead of coming across in the way of: "i don't care, this is the direction. Take it or fork it" (my interpretation of your response) you could try to explain it politely. Why the decision is made as it was and why this direction is so super important. Perhaps you did explain it in depth on other places, i don't know. If that's the case, then just link to it and be done with it :)

**Comment by liskin** [comment]
> there are just two things which put fonts on the screen in GTK apps: Pango and HarfBuzz. FreeType does some work as well but it's not fully utilized any longer. What does this leave us with? I'm confused.

Actually I don't think this is correct. HarfBuzz doesn't render. FreeType does the rendering, otherwise setting FREETYPE_PROPERTIES would have no effect, and it does, and not just to pango apps, but to all of them. Also, I think you missed cairo, and I'm still hopeful that maybe this is actually already fixed there and just not picked by distros yet.

**Comment by birdie-github** [comment]
> Actually I don't think this is correct. HarfBuzz doesn't render.

I was talking about getting kerning hints. Sorry for the confusion. And as far as I understand HarfBuzz is now responsible for that. Please correct me if I'm wrong again. I just report what I have problems with - I'm not sure I have to dig deep and understand the innards of the issue.

**Comment by khaledhosny** [comment]
Certain hinting modes can change glyph advance width, and this change depends on the pixel size. Pango previously was getting glyph advances from FreeType which support hinting. Pango then switched to using HarfBuzz for glyph advances because it wants size independent glyph advances. HarfBuzz does not support hinted glyph advances by design, and Pango is does not want them either. There is nothing to fix in HarfBuzz, if you are not agreeing with decisions made by Pango you need to discuss this in Pango issue tracker.

**Comment by behdad** [comment]
> Yeah, I understand that you've had a hidpi screen for a while and supporting lower resolutions seems boring.

Is that statement backed by facts or just feelings?

**Comment by behdad** [comment]
Anyone still screaming here, go fork Pango and exercise your Free Software "choice".  I haven't worked for Red Hat or maintained Pango for over ten years.  If you can't take the message, not my problem.  But you are shooting the messenger.

**Comment by liskin** [comment]
> Is that statement backed by facts or just feelings?

As I said before, I was just trying to understand your point of view. There may be other explanations, like you having a different taste, different visual acuity, using different fonts, … Having a HiDPI screen just seemed like the most plausible one. Sorry if I got that wrong. Feel free to tell us what's really behind this. :-)

**Comment by khaledhosny** [comment]
It had to be Phoronix garbage reporting after all! https://www.phoronix.com/scan.php?page=news_item&px=HarfBuzz-Hinting-Woe

**Comment by khaledhosny** [comment]
We probably should just lock this issue and move on.

**Comment by birdie-github** [comment]
@khaledhosny 

It's only _late_ in the discussion here it became obvious that the issue is not HarfBuzz' fault but how many people outside of HarfBuzz/Pango/Cairo/FreeType devevelopment teams really fully understand the intricacies of GTK font rendering? That's why I'm not sure if there's any garbage reporting, more like zero attention given to the [bug report](https://gitlab.gnome.org/GNOME/pango/issues/463) for almost two months.

Please close this bug report 'cause it surely looks like there's nothing left to discuss here.

**Comment by khaledhosny** [comment]
Phoronix has always been garbage content, clickpaity headlines and storm in a teacup kind of web site, nothing new here.

**Comment by uralets** [comment]
> Phoronix has always been garbage content, clickpaity headlines and storm in a teacup kind of web site, nothing new here.
ummm.. no. Also Michael has genuine leads and have posted exclusive content before it was made public officially on more than one occasion. Not to mention its practicality when you're out hunting for the next upgrade for your Linux box.
and if HIS headlines are clickbaity then.. I dunno, do you even open Internet?)

PS I'm not necessarily agreeing with the article in Q as far as its content and/or subject matter goes.

**Comment by khaledhosny** [comment]
Yes, by completely misinterpreting git commit messages and making scenes out of it, storm in a teacup indeed.

**Comment by liskin** [comment]
Well at least he saves us the time to find those git commits. As with any reporting, whether online or paper news, a massive amount of critical thinking is required to get the real truth out of it.

(But now we really do digress a lot. Sorry.)

**Comment by uralets** [comment]
> Yes, by completely misinterpreting git commit messages and making scenes out of it, storm in a teacup indeed.

Well in this case the headline is fairly correct and doesn't blame anyone. Indeed the reference to a HB "bug" responsible for the situ is wrong _as this and other discussions have shown_
I'll mention it to him on the forum.

Other than that the article seems OK?

**Comment by behdad** [comment]
> Feel free to tell us what's really behind this. :-)

Sure I'll give it a *brief* try here.

I stepped up maintaining Pango when it became unmaintained around 2004/2005.  In 2006 Red Hat hired me to keep working on that.  Around 2007 or so fontconfig also became unmaintained, so I picked that one as well.  I focused on making sure most of the planet's humans can *read at all* on their Free Software platforms.  That turned into what we know as HarfBuzz.

While my manager (Jonathan Blandford) and tech lead (Matthias Clasen) were supportive of my work on HarfBuzz and maintaining the text stack, the company and team priorities made it hard for them to keep staffing all of my time on these projects, so I was borrowed to different projects on the desktop team and that distracted me from my passion.

Eventually, in 2010 I moved to Chrome team at Google to continue working on HarfBuzz.  But again, my manager wanted me to spend time on other parts of the product as well, so I shifted to Google's Internationalization team in 2012 and enjoyed a long while of focusing on finishing HarfBuzz.  Since then HarfBuzz got adopted by Firefox, Chrome, Android, OpenJDK, LibreOffice, Adobe Photoshop / Illustrator, etc.  I made sure at least 5 billion people can read their preferred writing system on all those platforms.  Linux desktop benefited from all that effort nonetheless.

Pango became unmaintained when I left.  Cairo was unmaintained since about the same time (2010) and *still* is.  Matthias was maintaining GTK+ and Glib while also holding technical leadership at Red Hat's Desktop team.  Over the years many people stepped up to help with GTK+ and Glib, but no one ever got into maintaining Pango.

In that time i also worked on color emoji in the Free Software stack.  I posted a proof of concept based on cairo in 2013.  That was left unfinished because no one picked up finishing that work.  That was not interesting to my employer, and it wasn't work I was willing to do either, given how unmaintained cairo was (and is!).  It left Loud Linux Users complaining for many many years.  It wasn't until 2017 when Matthias and I met at GUADEC and I advised him about how to finish my work, and he did:

  https://blogs.gnome.org/mclasen/2017/08/18/post-guadec-distractions/

All of this while Xlib was being replaced by XCB, and later X replaced by Wayland, yet cairo was stagnated.  Just getting a release with new code from master was not happening.  No one with the capability was stepping up to the job.  Many tried.  It was an impossible task for a volunteer, and Red Hat didn't have resources to find someone and force them to maintain cairo.

At Google, I started leading the development of the next generation of Noto fonts.  Noto is the default font on Android for all scripts not served by the Latin/Greek/Cryillic Roboto.  While working on that I figured that we need to revive some old ideas from Apple/Adobe to be able to ship all of Noto on Android and web in sensible file sizes.  That work became OpenType 1.8 Font Variations, an industry-wide collaboration between Google, Microsoft, Apple, Adobe, and others.  To do that, I also led the building of a fully Open Source font compiler, known as fontmake, which is also adopted widely across the industry these days.

Again, I demoed variable-fonts on Pango at GUADEC in probably 2015.  Most of the changes went into FreeType & fontconfig, and some cairo just fine.  Exposing the APIs to Pango however was left undone since no one was actively working on Pango, and exposing all new API through Pango's FontMap abstraction, and then just leaving them unimplemented for non-FreeType backends (Windows, Mac) was already too much cruft work no one wanted to do.

Add to that the fact that Pango's Windows backend has been pretty much unmaintained for over 10 years as well, so is it's OS X backend.  All that stuff is unmaintained and undeveloped.

While all this happening, the world shifted to mobile phones.  Browsers became by far the biggest text renderer.  Microsoft itself dropped subpixel-rendering and GDI-style rendering in Windows 8; they had to bring some of it back by user pressure.  But that's 2012 we are talking about.  OS X has always had what people in this thread call blurry and unacceptable text.  GTK+ itself became pretty irrelevant in the mobile race after Nokia folded and Android took over.

Matthias and I managed to meet 2 / 3 times each year since 2018, trying to get some of the HarfBuzz advances into Pango.  All this while GTK4 was already switching to OpenGL-based rendering, making assumptions about text scaling linearly, an assumption that is at the core of this issue.

The challenge between the pixelated rendering vs scalable layout is nothing new.  Here's a treatise from almost 20 years ago: http://rastertragedy.com/ Here's my writeup about some of those same challenges I had to find solution to for Chrome back in 2012: http://goo.gl/yf3M7

LibreOffice switched to HarfBuzz to benefit from a single unified shaping-engine, same decision that Chrome and Firefox took as well.  These are engineering decisions.  It also meant LibreOffice dropped support for Type1 fonts.  There were of course the vocal minority complaining and wanting old features back.  I asked Adobe if they are willing to contribute support for Type1 to HarfBuzz, they said even Adobe products are dropping Type1 support next year.  Same thing about bitmap fonts, people want their bitmap fonts back, we told them to wrap them in OpenType containers.

And same change arrived at Pango.   Again, loud minority roars, throwing tantrums and unwilling to listen.  The world moves forward.  We do what we can with our limited resources.  We cannot keep your 20-year old favorite feature working forever.  It's not on us.  Find someone to maintain it and pay them to do for you if you cannot get over it.

When GNOME3 happened, enough people didn't like it for someone to put their money / time where their mouth was and forked GNOME2 and called it Mate or Cinnamon or whatever they liked.  I encourage that.   In the Pango bug I see someone saying now how Ubuntu, ArchLinux, and OpenSuse all default to hintfull.  Guess how many people from those distros *ever* contributed to the text rendering stack underlying GNOME?  ZERO.

And here's why I don't have much sympathy for the pixelated loud minority: everyone who has any expertise in this field agrees that this is very subjective.  I bet if those people use the "blurry" rendering for six months they forget they used something different ever.  Six month is an exaggeration.  I personally think people get used to it within two weeks.  But when they're using hyperbole calling Fedora "completely unusable" and whatever, there's no point engaging in a constructive conversation.  They only reason I jumped in past few days was to save our limited developer resources being wasted on replying to them.

So, here you go.  A *very brief* explanation.  Explaining takes time, which is the very problem at the root of this issue.  So asking us to explain is just asking us to give you what we don't have to begin with.  You want to help, sure, spend years learning and helping and one day you get to drive were things go.  Until then, the best you can do is to trust developers are doing the best they can.  Tell us what doesn't work for you, absolutely.  Harass us because you didn't get what you want?  I have no empathy for that.

**Comment by davelab6** [comment]
>> ![image](https://user-images.githubusercontent.com/833473/81475670-ebc01980-9222-11ea-8124-a7a8c3e71015.png)
>
> when they're using hyperbole calling Fedora "completely unusable" and whatever

I understand some users call non hintfull rendering unusable, but the menu shown above is **not the same** as the "blurry" rendering. 

Blurry rendering is something that many experts (including myself) believe is superior to hintfull, but only once you get through the temporary period of negative experience that is almost entirely down to what psychologists call "change aversion." To make the switch is like ripping a bandaid. Painful immediately and quickly subsiding. 

Sadly, the above menu text rendering is "keeping the band aid on with stick tape" - it renders fully hinted glyph shapes, but the "Metrics and Kerning" is not the same as before, because now those are coming from harfbuzz. 

(Which is why it's sort of correct to file the bug here, because indeed the issue is partially due to the introduction of harfbuzz...)

But, the root cause of the issue is trying to render fully hinted glyphs with unhinted metrics, and the only immediately viable solution is to stop trying to render the full hinting. 

As behdad has well explained, a longer term solution is to organize maintaining a legacy fork, like mate/cinnamon, either through direct action it by raising money and hiring someone. 

But the cost of doing that is probably more than the cost of buying hidpi kit for anyone who is still unhappy after ripping off the band-aid :)

**Comment by khaledhosny** [comment]
> (Which is why it's sort of correct to file the bug here, because indeed the issue is partially due to the introduction of harfbuzz...)

Nope, HarfBuzz does not control rendering. Pango cal always override the requested hinting mode to one that does not change horizontal metrics (which I suggested a while ago already) since it is not interested in hinted metrics anymore. That is a Pango issue not HarfBuzz one.

**Comment by davelab6** [comment]
Sure, but, harfbuzz is what needs to be overridden, and this is where actual answers have turned up ;)

**Comment by khaledhosny** [comment]
Cairo is where the rendering happens, and Cairo calls FreeType. Pango needs to tell Cairo to not use hintfull rendering mode. It doesn’t help when users also use the old FreeType rendering engine, AFAIK the new (and default) one ignores horizontal hinting as well.

**Comment by qwertychouskie** [comment]
Didn't have time to read through every post on every one of the many threads about this issue, but just wanted to say this:  Noto Sans 9 with slight hinting, v40 font renderer (the default nowadays), and RGB antialiasing still looks quite great, even on Ubuntu 20.04, which has the Pango version that breaks full hinting, even on a 1366x768 extra-low-DPI screen.  Y'all have options, the world is not imploding just yet ;)

**Comment by birdie-github** [comment]
> Cairo is where the rendering happens, and Cairo calls FreeType. Pango needs to tell Cairo to not use hintfull rendering mode. It doesn’t help when users also use the old FreeType rendering engine, AFAIK the new (and default) one ignores horizontal hinting as well.

I'm sorry I'm probably asking the most stupid question ever asked, but why do we need five different libraries for putting glyphs on the screen? freetype, fontconfig, cairo, harfbuzz and pango (not to mention some extra libraries for X11/Wayland). Why is it all so extremely complicated? It almost looks like someone intentionally made it this way. Somehow in the past we needed just three and they seemingly worked just fine: FreeType, fontconfig and libXft. What am I missing? Is this related to Wayland? Web browsers? Mobile OSes?

I'd also be grateful if someone explained in details why `hintstyle=hintfull` is something that absolutely needs to go and for which technical reasons aside from some future developments in regard to people switching to HiDPI displays which is not really occurring at any discernible rate as most popular computer monitors are 23-24" displays with a FullHD resolution and we still plenty of 14-15" laptops with a 1366x768 resolution with an atrocious pixel density. In short, a majority of monitors being sold even in 2020 are low pixel displays which do benefit greatly from subpixel rendering.

Lastly, as far as I can see subpixel rendering (ClearType v2 style) is very much alive in Windows 10 and is used by close to 100% of classic Win32 applications and to add insult to injury Microsoft has recently reneged on UWP and Win32 is very much alive.

**Comment by davelab6** [comment]
> In short, a majority of monitors being sold even in 2020 are low pixel displays which do benefit greatly from subpixel rendering.

But we are talking not about "blurry" subpixel rendering, but about "full hinting" (black and white) rendering.

**Comment by davelab6** [comment]
> slight hinting, v40 font renderer (the default nowadays), and RGB antialiasing still looks quite great

Exactly. This renders letterforms more as they were drawn and intended. Full hinting was a band aid needed in the late 80s to the early 00s, that distorted letterforms to fit the non color displays and then the CPUs of that era. 

That era is a far away now as mainframes were then.

**Comment by simoncozens** [comment]
>  Why is it all so extremely complicated?... Somehow in the past we needed just three and they seemingly worked just fine: FreeType, fontconfig and libXft. What am I missing?

Five billion people.

**Comment by xpue** [comment]
>if those people use the "blurry" rendering for six months they forget they used something different ever

I don't want to come off as rude, but this does not prove anything. With time, your perception can even get used to blurry image simply up-scaled from lower resolution.

**Comment by vvrein** [comment]
Hi everyone!
I'm on Arch Linux, I'm regularly update all my pkgs, and I don't have any issue with kerning and/or blurring...
WM - latest sway

<details><summary>pacman -Qi pango harfbuzz</summary>


Name            : pango
Version         : 1:1.44.7+11+g73b46b04-1
Description     : A library for layout and rendering of text
Architecture    : x86_64
URL             : https://www.pango.org/
Licenses        : LGPL
Groups          : None
Provides        : libpango-1.0.so=0-64  libpangocairo-1.0.so=0-64  libpangoft2-1.0.so=0-64  libpangoxft-1.0.so=0-64
Depends On      : libthai  cairo  libxft  harfbuzz  fribidi
Optional Deps   : None
Required By     : dunst  grim  gst-plugins-bad  gst-plugins-base  gtk2  gtk3  imv  librsvg  mako  pangomm  sway
Optional For    : imagemagick  suil
Conflicts With  : None
Replaces        : None
Installed Size  : 4.36 MiB
Packager        : Jan Alexander Steffens (heftig) <jan.steffens@gmail.com>
Build Date      : Tue 11 Feb 2020 05:20:58 EET
Install Date    : Wed 12 Feb 2020 04:47:59 EET
Install Reason  : Explicitly installed
Install Script  : No
Validated By    : Signature

Name            : harfbuzz
Version         : 2.6.5-2
Description     : OpenType text shaping engine
Architecture    : x86_64
URL             : https://www.freedesktop.org/wiki/Software/HarfBuzz
Licenses        : MIT
Groups          : None
Provides        : libharfbuzz.so=0-64  libharfbuzz-subset.so=0-64  libharfbuzz-gobject.so=0-64
Depends On      : glib2  freetype2  graphite  libglib-2.0.so=0-64  libfreetype.so=6-64  libgobject-2.0.so=0-64
Optional Deps   : cairo: hb-view program [installed]
Required By     : chromium  freetype2  harfbuzz-icu  libass  pango
Optional For    : None
Conflicts With  : None
Replaces        : None
Installed Size  : 4.61 MiB
Packager        : Evangelos Foutras <evangelos@foutrelis.com>
Build Date      : Mon 27 Apr 2020 13:21:18 EEST
Install Date    : Sat 02 May 2020 15:02:36 EEST
Install Reason  : Installed as a dependency for another package
Install Script  : No
Validated By    : Signature
</details>


<details><summary>And here how I see text</summary>

## Firefox:   
![image](https://user-images.githubusercontent.com/17022100/81494503-f4533700-9298-11ea-8cf3-07af3c990528.png)

## ReText:  
![image](https://user-images.githubusercontent.com/17022100/81494559-4b590c00-9299-11ea-80db-7743e83f1d9f.png)

</details>


So, idkw, but seems issue with fonts does not affect sway, or already fixed in pango (?).
Any ideas why all is ok? And, maybe users with font issues worth to try wayland?

**Comment by nim-nim** [comment]
> > Cairo is where the rendering happens, and Cairo calls FreeType. Pango needs to tell Cairo to not use hintfull rendering mode. It doesn’t help when users also use the old FreeType rendering engine, AFAIK the new (and default) one ignores horizontal hinting as well.
> 
> I'm sorry I'm probably asking the most stupid question ever asked, but why do we need five different libraries for putting glyphs on the screen? freetype, fontconfig, cairo, harfbuzz and pango (not to mention some extra libraries for X11/Wayland). Why is it all so extremely complicated? 

That’s because human text is an exceedingly complex subject. Unicode (1991, now in version 13) *still* has not finished identifying all the symbols real human text uses (let alone defined how to shape them). OpenType (1997, now in version 1.8.3) *still* has not finished mapping all the ways people want to represent those symbols on digital screens. And pre-HDPi screens just do not have enough pixels to render all the above at human text size, except via exceedingly complex heuristics (an heuristic is something that works most of the time and fails horribly the rest of the time).

That makes maintaining any modern font stack (as opposed to the extremely basic ASCII bitmap engines of last century) a titanic task. There are very few of them out there, be it on free desktop systems or elsewhere.

When people started working on this free desktop side in the 2000’s, they quickly realised it just could not fit within the existing software architecture. They wasted around a decade trying to fit it in anyway, before taking radical decisions like killing X11 and agreeing to converge on a single set of text rendering libs (*text* *rendering* is what killed X11 in the first place, no video or 3D games or whatever).
https://www.freedesktop.org/wiki/TextLayout2007/

Because of the huge amount of work that represented, and because it’s easier to get people to fund rewriting the login screen 20 times than work on core low-level software infrastructure,
this master plan is *still* being implemented. Pango is switching to Harbuzz just now, LibreOffice did it not so long ago.

In the meanwhile the inevitable complexity of the new text stack, is compounded by the existence of a huge amount of pre-2000 text libraries and font formats, which linger everywhere waiting for someone to replace them with the new text rendering architecture.

Even a wealthy desktop-oriented software company like Microsoft has not finished converting all the Windows software universe to its own modern text engine. Scratch a little and you find the same pervasive use of several layers of legacy text libs with the associated rendering problems. Even a killer application like Microsoft Office still selects fonts the pre-2000 GDI RBBI way, aka the legacy OpenType Name ID 1 & 2 layer (at least we got that fixed in most Free Desktop apps).

**Comment by davelab6** [comment]
> I don't have any issue with kerning and/or blurring...

Well that's good, because you are seeing the "blurry" rendering.

**Comment by davelab6** [comment]
> > if those people use the "blurry" rendering for six months they forget they used something different ever
> 
> I don't want to come off as rude, but this does not prove anything. With time, your perception can even get used to blurry image simply upscaled from lower resolution.

What see you trying to prove, then?

**Comment by xpue** [comment]
Continuing on with my example, "image upscaled from lower resolution" is of objectively lower quality.

**Comment by Monsterovich** [comment]
> You are using Free Software you realize, right?

Free Software or not, you must realize, your font rendering sucks ass now.

**Comment by twardoch** [comment]
@behdad A few people referred this thread to me, citing that in contains your “autobiography”. For the sake of completeness, I’ll add that in addition to “finishing HarfBuzz”, you also “finished fontTools”, a once-great, then-unmaintained Python library for font building, parsing and manipulation originated by Just van Rossum. Once a hacking and patching tool used by few, it now is the de facto standard for variable and static OpenType font creation (FontLab 7 uses it for the former), supports all of OpenType and includes 4x more functionality as before. 

I think it’s fair to say that with your work and leadership on HarfBuzz, fontTools, Noto and OT Variations, you’ve done more for fonts, text and, well, global literacy in a few years than many teams have done over a few decades. 💪👍

**Comment by markg85** [comment]
I'm getting really confused by the term "blurry" in the contexts given here. It could be the screenshot from @ebraminio yet that looks like something more is wrong.

I personally use slight RGB subpixel hinting on the noto fonts. Is that the "blurry" definitiion?

**Comment by ebraminio** [comment]
> It could be the screenshot from @ebraminio yet that looks like something more is wrong.

Please make another screenshot with your regular [not tweaked setup] and use that for comparison as that wasn't necessarily what end users get and is more to demonstrate shaping than rendering.

**Comment by lockywolf** [comment]
>>That’s because human text is an exceedingly complex subject. Unicode (1991, now in version 13) still has not finished identifying all the symbols real human text uses (let alone defined how to shape them). OpenType (1997, now in version 1.8.3) still has not finished

I in time when Serbia de facto dropped Cyrillic in favour of Latin, Kazakhstan dropped Cyrillic in favour of Latin, everyone younger than 50 is typing Chinese using Latin-based input method (pīnyīn), there are fast-and-efficient everything2Latin extensions for Chrome and Firefox (and an efficient latinification of every language imaginable) and there is a realistic hope that all the huge nightmare of legacy writing systems can be dropped and go to the same direction where cuneiform went, this work is not just titanical, it's also the most wasteful use of brainpower ever imagined.

Also, as lithographically-manufactured LEDs are extremely cheap, the proportion of low-dpi displays is not only not going down, it will be going up. Not in laptops, perhaps.

**Comment by stalkerg** [comment]
Stop chewing snot! Who wants to fork? Just do it.

In my opinion, is a communication failure with the community. Instead of asking the community about the important decision and ask help you just do it. Because this decision can affect millions of people you should be asked Michale about announce especially if this important library unmaintain. 
In any case, it's a shame and we should learn this lesson. 
Again Free Software does not mean developers can do what they want, it means community drive what the want.

**Comment by behdad** [comment]
> In my opinion, is a communication failure with the community. Instead of asking the community about the important decision and ask help you just did it. Because this decision can affect millions of people you should be asked Michale about announce especially if this important library unmaintain.
> In any case, it's a shame and we should learn this lesson.

Like these?

https://blogs.gnome.org/mclasen/2019/05/25/pango-future-directions/
https://blogs.gnome.org/mclasen/2019/07/19/pango-updates/

![image](https://user-images.githubusercontent.com/142891/81504185-b1935e00-929c-11ea-94bf-c12d5893ed82.png)

**Comment by markg85** [comment]
> > In my opinion, is a communication failure with the community. Instead of asking the community about the important decision and ask help you just did it. Because this decision can affect millions of people you should be asked Michale about announce especially if this important library unmaintain.
> > In any case, it's a shame and we should learn this lesson.
> 
> Like these?
> 
> https://blogs.gnome.org/mclasen/2019/05/25/pango-future-directions/
> https://blogs.gnome.org/mclasen/2019/07/19/pango-updates/
> 
> ![image](https://user-images.githubusercontent.com/142891/81504185-b1935e00-929c-11ea-94bf-c12d5893ed82.png)

That image does show you're using a high DPI screen ;)
Which - for some others in these comments - might make your opinion feel biased towards high DPI and not towards the still fairly regular monitors of say Full HD (1920x1080) around 24 inches. Or 1920x1200 at around the same inches.

At first, i had the impression you're resistance was to promote the adaptation of high DPI. Which, on the linux **desktop** (when scaling is involved that is) is still a deeply sad thorough nightmare to get working properly when you're **also** in a dual screen setup. Even if that setup is 2 screens of the same specs. The irony is that this very statement is completely irrelevant on mobile (linux powered too) where most devices are high DPI these days and have been for a couple of years.

Anyhow. Back to **blurry**. What is blurry? Is that what i described earlier (slight RGB subpixel hinting)?

**Comment by mattst88** [comment]
> Again Free Software does not mean developers can do what they want, it means community drive what they want.

This is simply false. The people that do the work make the decisions.

I often tell people that if they want to influence the decision making they can:

1. Do the work
2. Pay someone to do the work
3. Ask someone nicely to do the work

Outside of those options, there's only whining, which does not endear others to your cause.

**Comment by nim-nim** [comment]
> I have never even encountered a HiDPI display in the wild yet, certainly none that I can afford.

There are quite cheap today, both in laptop and standalone form, but you have to choose pixel density over other factors, instead of optimizing all the other aspects of your hardware and expecting software to perfrom text rendering miracles on crap screens. Lots of UHD monitors around 300/350 $ on amazon.com for example. That’s cheaper than a lot of smartphones, not much more expensive than entry level FHD monitors, and cheaper than a lot of high-end FHD monitors.

For laptop screens it’s a tad harder, but you just have to filter by screen resolution first instead of filtering by CPU first. The laptop screen represents an insane amount of the laptop price anyway, more than the CPU or memory, it’s idiotic to evaluate a laptop without looking at its most expensive component first (manufacturers will replace a broken laptop motherboard. If you break the screen, they won’t bother and change the whole thing).

**Comment by nim-nim** [comment]
> Anyhow. Back to **blurry**. What is blurry? Is that what i described earlier (slight RGB subpixel hinting)?

There is a school of users which has not gotten over Windows moving to scalable fonts yet, and insist on replicating 1990s bitmap rendering on modern hardware. They will call any text rendering that does not transform glyphs in anaemic horribly distorted shapes, with irregular metrics, as little black pixels as possible and no gray/color pixel at all blurry.

Because they are stuck in the 1990s they are also generally fans of the first generation of Microsoft scalable fonts, Tahoma & co (The “Core fonts for the web” Microsoft used to kill Netscape and lost interest in afterwards). Since it was a first attempt at hinted scalable fonts it was rather close to the (then current) bitmapy look. However, because it was a first gen attempt, it is full of dubious decisions which were removed in later fonts & text renderers.

**Comment by markg85** [comment]
> If you buy something with lower pixel density, well that’s something to look videos or play games upon, don’t expect it to render text well fine, it’s not the right hardware for text. A decade or two ago, when TVs went digital, people bought enormous TVs, tried to use them as computer screens, and complained text looked like crap video subtitles.

I'm glad your opinion is just that. Your opinion. As that is total bs!
The world (in desktop usage) is still littered with, say, ~96DPI monitors. That resolution is fine and text **must** render just fine on that. That has **nothing** to do with people stuck in 1990s (i'm from '85, i'm definitely not stuck in the 90').

The way you present it you need to have a display for games/videos (there you're 1920x1080 screens would fit) and a screen for "high quality" stuff. That's not how things work and you know that very well as an expert in fonts.

If you "force" 4k monitors down everyone's throat, they are impacted heavily when wanting to play a game. As 4k is 4x the pixels and that is quite taxing for even today's GPU's. Or are you also suggesting that everyone should be forced to having a 2080TI card?...

To continue the 4k insanity with a practical example. I'm writing this message right now on a 4k 28" monitor as that 4k resolution. Everything is rather small at that resolution. So what would your expert opinion be to fix it? Probably to turn on 2x scaling. Everything then becomes twice as big and - imho - just too big for the screen size. So i don't want that. Then you can still do fractal scaling. Say 1.5x. That - size wise - looks just fine. But text wise... ohh boy, everything really is ugly as hell!

**Comment by liskin** [comment]
> There is a school of users which has not gotten over Windows moving to scalable fonts yet

At certain very small sizes it really is better... Or is that subjective as well? I don't know, maybe it is.

Anyway, to prove my point, here are two picture of a terminal with a very small font, both using DejaVu Sans Mono, with/without antialiasing/subpixel rendering.
<details>
<img src="https://user-images.githubusercontent.com/300342/81508593-3b771180-9305-11ea-8cfe-e24bbb57ad7f.png">
<img src="https://user-images.githubusercontent.com/300342/81508596-3e720200-9305-11ea-9ce0-6a1f9a1cf68c.png">
</details>

With the classic Fixed font, one can go even smaller (yes, I can still read that on a fullhd 14" laptop):
<details>
<img src="https://user-images.githubusercontent.com/300342/81508608-577ab300-9305-11ea-8319-6e7679b7ea03.png">
</details>

That being said, I decided that it really will be the best option to switch to slight hinting, v40 freetype rasterizer and pango 1.44 and just get used to it. Took me a while to get subpixel positioning working in gtk, but it now does and it's tolerable. Firefox is unfortunately still a bit broken (uses greyscale antialiasing in many UI elements).

So, um, I can't say I'm impressed, but there really isn't a better way going forward, is it?

**Comment by vvrein** [comment]
My video about current font situation made with logitech c170.
https://tort.icu/bin/font.mp4  

With 
<details><summary>fonts.conf</summary>

```
<?xml version='1.0'?>
<!DOCTYPE fontconfig SYSTEM 'fonts.dtd'>
<fontconfig>
	<match>
		<edit mode="prepend" name="family"><string>Noto Sans</string></edit>
	</match>
	<match target="pattern">
		<test qual="any" name="family"><string>serif</string></test>
		<edit name="family" mode="assign" binding="same"><string>Noto Serif</string></edit>
	</match>
	<match target="pattern">
		<test qual="any" name="family"><string>sans-serif</string></test>
		<edit name="family" mode="assign" binding="same"><string>Noto Sans</string></edit>
	</match>
	<match target="pattern">
		<test qual="any" name="family"><string>monospace</string></test>
		<edit name="family" mode="assign" binding="same"><string>Noto Mono</string></edit>
	</match>
	<match target="font">
		<edit mode="assign" name="antialias">
			<bool>true</bool>
		</edit>
		<edit mode="assign" name="hinting">
			<bool>true</bool>
		</edit>
		<edit mode="assign" name="hintstyle">
			<const>hintslight</const>
		</edit>
		<edit mode="assign" name="lcdfilter">
			<const>lcddefault</const>
		</edit>
		<edit mode="assign" name="rgba">
			<const>rgb</const>
		</edit>
		<edit mode="assign" name="embeddedbitmap">
			<bool>true</bool>
		</edit>
	</match>

</fontconfig>
```

</details>

And `FREETYPE_PROPERTIES=truetype:interpreter-version=40`
And latest pango & harfbuzz & whatever there are more.

Is it blurry?  Is there anything wrong with this?  

I'm trying understand here, what's wrong with text, or why people are so angry to @behdad.

Has anyone tried to ask MS to make anything with fonts? I'm not.
But I'm sure you will get a well-designed and beautiful speech about GTFO.
Same thing here in OSS|FS , but with clear GTFO and for free :)

**Comment by stalkerg** [comment]
>Like these?

It's excellent illustration of my words. All this is one way communication.

**Comment by behdad** [comment]
> > Like these?
> 
> It's excellent illustration of my words. All this is one way communication.

Time for you to go home.

**Comment by nim-nim** [comment]
> I'm glad your opinion is just that. Your opinion. As that is total bs!
> The world (in desktop usage) is still littered with, say, ~96DPI monitors. That resolution is fine and text **must** render just fine on that. 

That’s physics, not an opinion. Representing shapes on a discrete grid is a form of sampling. Sampling frequency limits the accuracy at which the original non-discrete signal can be reproduced accurately (the best known proof of that is the Shannon–Hartley theorem). Hinting is attempt to limit the effect of physics by playing with the way the human eye works, it can not make physics go away and it is only damage limitation.

> If you "force" 4k monitors down everyone's throat, they are impacted heavily when wanting to play a game. As 4k is 4x the pixels and that is quite taxing for even today's GPU's. 

So, just as I wrote, there is no cheap HiDPI screen supply problem, there is a class of users that deliberately buy screens optimized for other usages, than text rendering.

**Comment by rwyrz** [comment]
> > > That’s because human text is an exceedingly complex subject. Unicode (1991, now in version 13) still has not finished identifying all the symbols real human text uses (let alone defined how to shape them). OpenType (1997, now in version 1.8.3) still has not finished
> 
> I in time when Serbia de facto dropped Cyrillic in favour of Latin, Kazakhstan dropped Cyrillic in favour of Latin, everyone younger than 50 is typing Chinese using Latin-based input method (pīnyīn), there are fast-and-efficient everything2Latin extensions for Chrome and Firefox (and an efficient latinification of every language imaginable) and there is a realistic hope that all the huge nightmare of legacy writing systems can be dropped and go to the same direction where cuneiform went, this work is not just titanical, it's also the most wasteful use of brainpower ever imagined.
> 
> Also, as lithographically-manufactured LEDs are extremely cheap, the proportion of low-dpi displays is not only not going down, it will be going up. Not in laptops, perhaps.

If you think "there is a realistic hope that all the huge nightmare of legacy writing systems can be dropped" you've got another thing coming.  
For one, people are still making new ones, like Adlam.

Two, people identify themselves with them - some even go so far as to call roughly the same language different names depending on which script it's written in (to be fair, there usually are some differences in preferred vocabulary, as well), in fact.

Three, what you call "efficient latinification for every language" is, more often than not, actually a quick-and-dirty kludge created for the benefit of English speakers and computer systems that don't understand Unicode that drops phonemic distinctions considered 'unimportant' much like was done in, say, ancient Greece. Or are you talking about systems used mostly in universities and other scientific publications?

Four, I find your examples somewhat questionable. Kazakhstan doesn't seem anywhere close to adopting the Latin script for real. Nazarbayev's alphabet has been revised twice so far and it doesn't seem like any government websites use it, either. As for pinyin input, you're ignoring Hong Kong and Taiwan, where the defaults are Cangjie and Zhuyin, respectively, neither of which is Latin transcription. Especially in Hong Kong, I doubt "everyone under 50" would use pinyin input, because you can't really type Cantonese properly with that.

Five, Latin is not a good fit for every language. For example, Japanese does a lot with kanji that a purely phonetic writing system simply can't. Distinguishing homophones, emphasizing different meanings of the same word, using kanji glossed with completely different words in songs and other art forms

Six, do you really expect these scripts to go away? Remember that the Unicode Consortium is continually encoding more and more characters *no one* is using. Why? Because they occur in texts already written. And researchers want to be able to digitize those texts, and I imagine they want their computers to be able to render the effect graphically… Even if the whole world abandoned all the other scripts and went Latin exclusively, the burden of supporting all that wouldn't go away, you know. Because the can of worms has been opened already, new text is being created in tens, if not hundreds of different scripts as write this. Do you really think all that can be made to disappear? I certainly don't.

**Comment by liskin** [comment]
> So, just as I wrote, there is no cheap HiDPI screen supply problem, there is a class of users that deliberately buy screens optimized for other usages, than text rendering.

Yes, deliberately, because font rendering is not the only thing in the world. A laptop has many parts: keyboard, touchpad, possibly trackpoint, (external) battery (or two), etc. I, for example, primarily care about the input devices (keyboard, touchpad, trackpoint), and only then I look for a good display, therefore I'm stuck with a ThinkPad that happens to be one generation before they introduced eDP wide enough to connect a HiDPI display.

So yeah, if you're willing to sacrifice everything else to get crisp fonts, you can get a cheap HiDPI display. If you want to get a good overall experience with your laptop, it's not that easy, and that's why people resort to software workarounds.

**Comment by nim-nim** [comment]
> > So, just as I wrote, there is no cheap HiDPI screen supply problem, there is a class of users that deliberately buy screens optimized for other usages, than text rendering.
> 
> Yes, deliberately, because font rendering is not the only thing in the world. 

So what? You don’t expect a computer with a weak CPU to perform well on scientific computing do you? When you optimize away the screen parameter that makes it possible to render text at high detail level, don’t complain the result look bad at small text sizes.

**Comment by behdad** [comment]
Can you all please stay on topic?

**Comment by markg85** [comment]
> > If you "force" 4k monitors down everyone's throat, they are impacted heavily when wanting to play a game. As 4k is 4x the pixels and that is quite taxing for even today's GPU's.
> 
> So, just as I wrote, there is no cheap HiDPI screen supply problem, there is a class of users that deliberately buy screens optimized for other usages, than text rendering.

And then there seems to be a class of devs, like you, that don't give a damn and would eagerly force fugly font rendering down our throat to force us to get a HiDPI screen.

Now we didn't even get into the definition of HiDPI... Is that QHD on 28 inch? 4K on 28 inch? How much DPI (as that might be an easier metric).

I'm sorry to say, but your opinion is a utopian one that does not reflect the world as it is today nor will it be there in the next decade. I personally am all for cutting edge so i'd be must likely to pick your side in this. But even then i still find your opinion too wrong to be sensible. You **cannot** demand that the whole world is going to use HiDPI just because **you** (or the font developer community) finds that to be best. You also cannot demand the world to move in the direction to have monitors dedicated to specific workloads. That would suck immensely!

**Comment by nim-nim** [comment]
> Now we didn't even get into the definition of HiDPI... Is that QHD on 28 inch? 4K on 28 inch? How much DPI (as that might be an easier metric).

Just remember that computer prints were not considered good before the advent of laser printers, and those were introduced at 300dpi. (Practically, 96dpi is so bad, you start getting visible improvements of text accuracy around 150dpi).

**Comment by Szunti** [comment]
As I see Harfbuzz already supports hinting. Isn't it enough to call hb_ft_font_set_load_flags () with the correct hinting flags?

**Comment by khaledhosny** [comment]
> As I see Harfbuzz already supports hinting. Isn't it enough to call hb_ft_font_set_load_flags () with the correct hinting flags?

This requires using FreeType font functions for loading the fonts and getting glyph advances, which is what Pango moved away with. HarfBuzz’s own font loading functions don’t support hinting.

**Comment by davelab6** [comment]
> Back to **blurry**. What is blurry? Is that what i described earlier (slight RGB subpixel hinting)?

Yes

**Comment by baybal** [comment]
> > As I see Harfbuzz already supports hinting. Isn't it enough to call hb_ft_font_set_load_flags () with the correct hinting flags?
> 
> This requires using FreeType font functions for loading the fonts and getting glyph advances, which is what Pango moved away with. HarfBuzz’s own font loading functions don’t support hinting.

And this is what should be the actual focus on the discussion on the improvement. I believe @behdad will not be against at all if somebody will find a practical way to 1) marry hinting from fonts with all the intricacies of positioning, and shaping happening inside Harfbuzz, 2) write the loading code for hints. Am I right?

As for bitmap fonts, @behdad would you agree to include a bitmap font loader?

**Comment by behdad** [comment]
> As for bitmap fonts, @behdad would you agree to include a bitmap font loader?

Happy to discuss, yes. Which particular bitmap formats are we talking about?

**Comment by jfkthame** [comment]
> a bitmap font loader

@baybal It'd probably be best to open a new issue about this, rather than extending this one.

**Comment by baybal** [comment]
@behdad it did not come to me before that hb already had OTB support, and a converter available. Nevermind.

More interested to hear what you think if there will be a desire from somebody to carry work on making hb to be aware of hints from fonts.

**Comment by behdad** [comment]
> @behdad it did not come to me before that hb already had OTB support, and a converter available. Nevermind.

No worries.

> More interested to hear what you think if there will be a desire from somebody to carry work on making hb to be aware of hints from fonts.

HarfBuzz doesn't currently do rasterization. Implementing hinting in HarfBuzz while rasterization happens elsewhere does not make much sense. Clients should advise HarfBuzz (using existing API) to load glyph extents from the same place that hinted rasterization hapens.

**Comment by baybal** [comment]
> HarfBuzz doesn't currently do rasterization. Implementing hinting in HarfBuzz while rasterization happens elsewhere does not make much sense.

I imagine it being quite an inelegant, and platform dependent hack to provide feedback from hb<cairo<freetype, but how else to do it.

**Comment by behdad** [comment]
> I imagine it being quite an inelegant, and platform dependent hack to provide feedback from hb<cairo<freetype, but how else to do it.

What problem are you trying to solve?

**Comment by baybal** [comment]
> > I imagine it being quite an inelegant, and platform dependent hack to provide feedback from hb<cairo<freetype, but how else to do it.
> 
> What problem are you trying to solve?

Provide a working strong hinting support in HB for people who need it.

**Comment by khaledhosny** [comment]
> Provide a working strong hinting support in HB for people who need it.

Which by definition requires getting the advances from the rasterizer (no two rasterizers produce identical output, especially with the different options various rasterizers support). This is already supported by HarfBuzz, nothing is hacky about it.

**Comment by afontenot** [comment]
Can I ask a clarificatory question about this old bug?

Obviously the OP of this issue is not entitled to having pango / harfbuzz devs implement full hinting support for them. But I think they are right to point out that the actual result is unacceptably broken: compare [before, pango+freetype](https://bugzilla-attachments.redhat.com/attachment.cgi?id=1639774) to [after, pango+harfbuzz](https://bugzilla-attachments.redhat.com/attachment.cgi?id=1639436).

My reading of the situation is that full hinting is no longer supported with pango ≥ 1.44, because of the switch to using harfbuzz to provide kerning. However, all (?) common desktop environments still support setting hintfull, and many users will set it manually in their fontconfig files either not knowing that it is no longer supported in Pango or else importing a pre-existing config. Moreover, expecting users, desktop environment projects, and distributions to move away from using hintfull seems like a mistake. Many programs may still use fontconfig directly to do hinting, in which case hintfull continues to work. 

In other words the bug is this: Pango continues to attempt to use the hintfull setting from fontconfig even though it no longer supports horizontal hinting. Instead, Pango should reject hintfull, and use hintslight (or medium?) instead when it is detected. That way, hintfull will continue to work in programs that support it, and programs that use Pango will get the best output they can, at least until full hinting support is re-added (which may very well never happen).

Does this seem like a reasonable assessment of the issue? Does the latest version of Pango still generate the broken output indicated by the OP? Should a bug be filed with that project to adopt the behavior described above? I wasn't able to replicate the issue using pango-view with Pango 1.48, but I might be missing something.

Note: I'm not personally affected by this bug, I've always used hintnone because I strongly prefer it, but I ran across this issue while working on Pango stuff recently and was surprised to see such broken rendering.

**Comment by khaledhosny** [comment]
You are in the wrong bug tracker. Pango issues should be discussed in Pango bug tracker not HarfBuzz’s.

**Comment by afontenot** [comment]
> You are in the wrong bug tracker. Pango issues should be discussed in Pango bug tracker not HarfBuzz’s.

Yep, I figured as much, although it wasn't clear to me that this is definitely a Pango issue (no, I didn't read all 90 comments). Unfortunately the relevant Pango issue was locked because of the behavior of the OP, so most discussion has happened here. I was hoping to get some insight on whether this is still an issue before bringing it up there again, but maybe I'll just move on, since as I said, the issue doesn't affect me.

**Comment by birdie-github** [comment]
> Does this seem like a reasonable assessment of the issue?

Pango/Harfbuzz developers must know :-)

> Does the latest version of Pango still generate the broken output indicated by the OP?

Yes.

![bug](https://user-images.githubusercontent.com/4958603/141077597-8cb52861-13f7-47cb-9c91-77d0ddb07d65.png)

> Should a bug be filed with that project to adopt the behavior described above?

_"Pango issues should be discussed in Pango bug tracker not HarfBuzz’s."_
